pstoreds/addr_book_gc.go

```diff
@@ -0,0 +1,299 @@
+			if err := batch.Put(gcKey, []byte{}); err != nil {
+				log.Warningf("failed to add new GC key: %v, err: %v", gcKey, err)
+			}
+		}
+	}
+
+	results, err := ab.ds.Query(purgeQuery)
+	if err != nil {
+		log.Warningf("failed while fetching entries to purge: %v", err)
+		return
+	}
+	defer results.Close()
+
+	now := time.Now().Unix()
+
+	// keys: 	/peers/gc/addrs/<unix timestamp of next visit>/<peer ID b32>
+	// values: 	nil
+	for result := range results.Next() {
+		gcKey := ds.RawKey(result.Key)
+		ts, err := strconv.ParseInt(gcKey.Parent().Name(), 10, 64)
+		if err != nil {
+			dropInError(gcKey, err, "parsing timestamp")
+			log.Warningf("failed while parsing timestamp from key: %v, err: %v", result.Key, err)
+			continue
+		} else if ts > now {
+			// this is an ordered cursor; when we hit an entry with a timestamp beyond now, we can break.
+			break
+		}
+
+		idb32, err := b32.RawStdEncoding.DecodeString(gcKey.Name())
+		if err != nil {
+			dropInError(gcKey, err, "parsing peer ID")
+			log.Warningf("failed while parsing b32 peer ID from key: %v, err: %v", result.Key, err)
+			continue
+		}
+
+		id, err = peer.IDFromBytes(idb32)
+		if err != nil {
+			dropInError(gcKey, err, "decoding peer ID")
+			log.Warningf("failed while decoding peer ID from key: %v, err: %v", result.Key, err)
+			continue
+		}
+
+		// if the record is in cache, we clean it and flush it if necessary.
+		if e, ok := ab.cache.Peek(id); ok {
+			cached := e.(*addrsRecord)
+			cached.Lock()
+			if cached.Clean() {
+				if err = cached.Flush(batch); err != nil {
+					log.Warningf("failed to flush entry modified by GC for peer: &v, err: %v", id.Pretty(), err)
+				}
+			}
+			dropOrReschedule(gcKey, cached)
+			cached.Unlock()
+			continue
+		}
+
+		record.Reset()
+
+		// otherwise, fetch it from the store, clean it and flush it.
+		entryKey := addrBookBase.ChildString(gcKey.Name())
+		val, err := ab.ds.Get(entryKey)
+		if err != nil {
+			// captures all errors, including ErrNotFound.
+			dropInError(gcKey, err, "fetching entry")
+			continue
+		}
+		err = record.Unmarshal(val)
+		if err != nil {
+			dropInError(gcKey, err, "unmarshalling entry")
+			continue
+		}
+		if record.Clean() {
+			err = record.Flush(batch)
+			if err != nil {
+				log.Warningf("failed to flush entry modified by GC for peer: &v, err: %v", id.Pretty(), err)
+			}
+		}
+		dropOrReschedule(gcKey, record)
+	}
+
+	if err = batch.Commit(); err != nil {
+		log.Warningf("failed to commit GC purge batch: %v", err)
+	}
+}
+
+// populateLookahead populates the lookahead window by scanning the entire store and picking entries whose earliest
+// expiration falls within the new window.
+//
+// Those entries are stored in the lookahead region in the store, indexed by the timestamp when they need to be
+// visited, to facilitate temporal range scans.
+func (ab *dsAddrBook) populateLookahead() {
+	if !atomic.CompareAndSwapInt32(&ab.gcLookaheadRunning, 0, 1) {
```

Comment:
Same with these. We should be able to simply drop this guard.

Reply:
See https://github.com/libp2p/go-libp2p-peerstore/pull/47/#discussion_r241016934